File: CudaSimplex/SimplexSolver.CS/Dados/Restricao.cs
﻿
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;


namespace SimplexSolver.CS.Dados
{

  public class Restricao
  {

    public string Nome { get; set; }
    //Propriedades que compõem uma função de restrição
    public Dictionary<string, Variavel> Variaveis { get; set; }
    //Conjunto de variáveis
    public Desigualdade Desigualdade { get; set; }
    //Um tipo de relacionamento
    public double TermoLivre { get; set; }
    //Valor do termo livre
    public Variavel VariavelArtificial { get; set; }
    //Variavel artificial que deve ser adicionada para tornar a restrição uma igualdade
    public bool Normalizado { get; set; }
    //Apenas uma variável de controle para saber se a função de restrição está normalizada (forma padrão)

    public Restricao()
    {
      this.Variaveis = new Dictionary<string, Variavel>();
    }

    public Restricao(string nomeRest)
      : this()
    {
      Nome = nomeRest;
    }

    public void AddVariavel(string varName, double varValue)
    {
      //Verificar se ja existe essa variavel na restriçao
      Variavel varAux = null;

      if (Variaveis.ContainsKey(varName))
      {
        varAux = Variaveis[varName];
      }
      else
      {
        varAux = new Variavel();
        varAux.Nome = varName;
        Variaveis.Add(varName, varAux);
      }

      varAux.Coeficiente = varValue;
    }

    public override string ToString()
    {

      string restString = string.Empty;


      if (!Normalizado)
      {
        foreach (Variavel var in this.Variaveis.Values)
        {
          if (var.Coeficiente > 0)
          {
            restString += " + " + var.Coeficiente + var.Nome;
          }
          else
          {
            restString += var.Coeficiente + var.Nome;
          }
        }

        if (this.Desigualdade == Desigualdade.MaiorOuIgual)
        {
          restString += " >= ";
        }
        else
        {
          restString += " <= ";
        }

        restString += TermoLivre.ToString();

      }
      else
      {
        restString += VariavelArtificial.Nome + " = " + TermoLivre + " - (";

        foreach (Variavel var in this.Variaveis.Values)
        {
          if (var.Coeficiente > 0)
          {
            restString += " + " + var.Coeficiente + var.Nome;
          }
          else
          {
            restString += var.Coeficiente + var.Nome;
          }
        }

        restString += " ) ";
      }

      return restString;
    }

  }

}

File: SimplexSolver/SimplexSolver/Dados/Restricao.vb
Public Class Restricao

    Public Enum Relacionamento
        MenorIgual = 0
        MaiorIgual = 1
    End Enum

    'Propriedades que compõem uma função de restrição
    Property Variaveis As New Dictionary(Of String, Variavel) 'Conjunto de variáveis
    Property Relation As Relacionamento = Nothing 'Um tipo de relacionamento
    Property TermoLivre As Double 'Valor do termo livre
    Property VariavelArtificial As Variavel 'Variavel artificial que deve ser adicionada para tornar a restrição uma igualdade
    Property Normalizado As Boolean = False 'Apenas uma variável de controle para saber se a função de restrição está normalizada (forma padrão)


    Sub setVariavelValue(ByVal varName As String, ByVal varValue As Double)

        'Verificar se ja existe essa variavel na restriçao
        Dim varAux As Variavel = Nothing

        If Variaveis.ContainsKey(varName) Then
            varAux = Variaveis(varName)
        Else
            varAux = New Variavel
            varAux.Nome = varName
            Variaveis.Add(varName, varAux)
        End If

        varAux.Coeficiente = varValue
    End Sub

    Sub deleteVariavel()
        Variaveis.Remove("X" & Variaveis.Count)
    End Sub

    Public Overrides Function ToString() As String

        Dim restString As String = String.Empty

        If Not Normalizado Then

            For Each var As Variavel In Me.Variaveis.Values
                If var.Coeficiente > 0 Then
                    restString += " + " & var.Coeficiente & var.Nome
                Else
                    restString += var.Coeficiente & var.Nome
                End If
            Next

            If Me.Relation = Relacionamento.MaiorIgual Then
                restString += " >= "
            Else
                restString += " <= "
            End If

            restString += TermoLivre.ToString
        Else

            restString += VariavelArtificial.Nome & " = " & TermoLivre & " - ("

            For Each var As Variavel In Me.Variaveis.Values
                If var.Coeficiente > 0 Then
                    restString += " + " & var.Coeficiente & var.Nome
                Else
                    restString += var.Coeficiente & var.Nome
                End If
            Next

            restString += " ) "
        End If

        Return restString
    End Function

End Class
